DIPs/DIP1xxx-RC.md

```diff
@@ -0,0 +1,232 @@
+
+Named arguments may be specified on structs, classes, unions, template blocks and mixin templates. As well as functions and methods. When used with structs, classes, unions or template blocks named arguments may be accessed by their identifier. E.g.
+
+```D
+struct MyWorld<string Name> {
+}
+
+static assert(MyWorld!(Name:"Earth").Name) == "Earth");
+
+void hello(<string message="World!">) {
+	import std.stdio;
+	writeln("Hello ", message);
+}
+
+void main() {
+	// Will print "Hello Walter"
+	hello(message : "Walter");
+}
+```
+
+This feature is quite convenient because it means that variadic template arguments and variadic function arguments can come before named arguments. So:
+
+```D
+void func(T..., <alias FreeFunc>)(T t, <bool shouldFree=false>) {
+}
+
+void abc() {
+	func!(FreeFunc : someRandomFunc)(1, 2, 3, shouldFree : true);
+}
+```
+
+Is in fact valid.
+
+For convenience at the definition side, you may combine named arguments together or keep them seperate, it is up to you.
+
+```D
+struct TheWorld<string Name, Type> {
+}
+
+alias YourWorld = TheWorld!(Name:"Here", Type:size_t);
+
+void goodies(T, <T t>, <U:class=Object>)(string text) {
+}
+
+alias myGoodies = goodies!(int, t:8);
+alias myOtherGoodies = goodies!(U:Exception, string, t : "hi!");
+```
+
+Any symbol that has named arguments, the named arguments are not considered for overload resolution. This puts a requirement on the unnamed arguments being unique and easily resolved.
+
+### Grammar changes
+
+```diff
+TemplateParameters:
++    < NamedTemplateParameterList|opt >
+
+TemplateParameter:
++    < NamedTemplateParameterList|opt >
+
++ NamedTemplateParameterList:
++    NamedTemplateParameter
++    NamedTemplateParameter ,
++    NamedTemplateParameter , NamedTemplateParameterList
+
++ NamedTemplateParameter:
++    Identifier = TemplateParameter
++    alias Identifier
+
+TemplateArgument:
++   NamedTemplateArgumentList
+
++ NamedTemplateArgumentList:
++     NamedTemplateArgument
++     NamedTemplateArgument ,
++     NamedTemplateArgument , NamedTemplateArgumentList
+
++ NamedTemplateArgument:
++     Identifier : TemplateArgument
+
+Parameter:
++   < NamedParameterList|opt >
+
++ NamedParameterList:
++    Parameter
++    Parameter ,
++    Parameter , NamedParameterList
+
++ NamedArgument:
++    Identifier : ConditionalExpression
+
+TraitsKeyword:
++    getNamedParameters
+```
+
+### Use cases
+
+#### Ranges
+
+Removes the usage of ``ElementType`` and initiations of it by making the type used a member.
+
+```D
+struct Adder<SourceType, Type=SourceType.Type> {
+    SourceType source;
+    Type toAdd;
+
+    @property {
+        Type front() {
+            return source.front() + toAdd;
+        }
+        
+        bool empty() {
+            return source.empty();
+        }
+    }
+    
+    void popFront() {
+        source.popFront();
+    }
+}
+
+auto adder(Source)(Source source, Source.Type toAdd) {
+    return Adder!(SourceType: Source, Type: Source.Type)(source, toAdd);
+}

```

Comment:
```d
auto adder(Source)(Source source, Source toAdd) {
    return Adder!Source(source, toAdd);
}
```

Reply:
I'm sorry, but I don't see what is wrong by the existing example?

That example is meant to show how to specify and override defaults and the general flow of its usage in the form of input ranges. Your replacement ignores ``toAdd`` being a separate thing (think ``InputRange!int`` with ``ElementType`` giving ``int``).

Reply:
Oh sorry, my bad. I missed the idea here but now it makes sense.